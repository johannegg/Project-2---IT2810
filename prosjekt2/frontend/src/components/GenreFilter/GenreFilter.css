.filterContainer {
	margin: 0;
	padding: 10px;
	border-radius: 10px;
	box-shadow: 0 4px 8px 0 #00000033;
	text-align: left;
	display: inline-block;
	width: 200px;
	background-color: white;
}

.filterRow {
	display: flex;
	align-items: center;
}

.filterContainer h2 {
	margin: 0;
	padding: 0;
	font-size: 1.5em;
	font-weight: 500;
}

input[type="checkbox"] {
	accent-color: var(--theme-green);
	cursor: pointer;
	margin: 10px;
}

.filterHeader {
	display: flex;
	align-items: center;
}

.disabled-filter {
	opacity: 0.5;
	cursor: not-allowed;
}

.disabled-filter-label {
	color: var(--theme-light-grey);
	cursor: not-allowed;
}

.filterCount {
	color: var(--theme-grey);
}

.disabled-filter-label .filterCount {
	color: var(--theme-light-grey);
}

@media (max-width: 1050px) {
	.filterContainer {
		width: 70%;
	}

	input[type="checkbox"] {
		width: 16px;
		height: 16px;
	}
}

@media (max-width: 600px) {
	.filterContainer {
		width: 300px;
	}
}

@media (max-width: 400px) {
	.filterContainer {
		width: 190px;
	}
}

@media (prefers-color-scheme: dark) {
	.filterContainer {
		background-color: var(--theme-almost-black);
		color: white;
	}
}
