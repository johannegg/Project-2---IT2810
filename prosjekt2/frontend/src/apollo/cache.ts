import { InMemoryCache, makeVar } from "@apollo/client";
import { SongData } from "../utils/types/SongTypes";
import { PlaylistData } from "../pages/Playlists/Playlists";

// Initialize sessionStorage and localStorage if values are not already set
const initializeStorage = (key: string, defaultValue: string) => {
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, defaultValue);
  }
};

const initializeLocalStorage = (key: string, defaultValue: string) => {
  if (!localStorage.getItem(key)) {
    localStorage.setItem(key, defaultValue);
  }
};

// Initialize values in storage
initializeStorage("minViews", "0");
initializeStorage("maxViews", "1000000");
initializeStorage("selectedGenres", "[]");
initializeStorage("sortOption", "views_desc");
initializeStorage("homeSearchTerm", "");
initializeStorage("favoritesSearchTerm", "");
initializeLocalStorage("favoriteSongs", "[]");
initializeLocalStorage("playlists", "[]");

// Load saved values from storage
const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
const savedMinViews = Number(sessionStorage.getItem("minViews"));
const savedMaxViews = Number(sessionStorage.getItem("maxViews"));
const savedSortOption = sessionStorage.getItem("sortOption") || "views_desc";
const savedHomeSearchTerm = sessionStorage.getItem("homeSearchTerm") || "";
const savedFavoritesSearchTerm = sessionStorage.getItem("favoritesSearchTerm") || "";
const savedFavoriteSongs = JSON.parse(localStorage.getItem("favoriteSongs") || "[]");
const savedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");

// Create reactive variables
export const genreFilterVar = makeVar<string[]>(savedGenres);
export const sortOptionVar = makeVar<string>(savedSortOption);
export const minViewsVar = makeVar<number>(savedMinViews);
export const maxViewsVar = makeVar<number>(savedMaxViews);
export const homeSearchTermVar = makeVar<string>(savedHomeSearchTerm);
export const favoritesSearchTermVar = makeVar<string>(savedFavoritesSearchTerm);
export const favoriteSongsVar = makeVar<SongData[]>(savedFavoriteSongs);
export const playlistsVar = makeVar<PlaylistData[]>(savedPlaylists);  
export const songDataVar = makeVar<SongData[]>([]);

// Configure cache with reactive variables
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        genreFilter: {
          read() {
            return genreFilterVar();
          },
        },
        sortOption: {
          read() {
            return sortOptionVar();
          },
        },
        minViews: {
          read() {
            return minViewsVar();
          },
        },
        maxViews: {
          read() {
            return maxViewsVar();
          },
        },
        homeSearchTerm: {
          read() {
            return homeSearchTermVar();
          },
        },
        favoritesSearchTerm: {
          read() {
            return favoritesSearchTermVar();
          },
        },
        favoriteSongs: {
          read() {
            return favoriteSongsVar();
          },
        },
        playlists: {
          read() {
            return playlistsVar();
          },
        },
        songData: {
          read() {
            return songDataVar();
          },
        },
      },
    },
  },
});

// Sync reactive variables with sessionStorage and localStorage
homeSearchTermVar.onNextChange((newHomeSearchTerm) => {
  sessionStorage.setItem("homeSearchTerm", newHomeSearchTerm);
});

favoritesSearchTermVar.onNextChange((newFavoritesSearchTerm) => {
  sessionStorage.setItem("favoritesSearchTerm", newFavoritesSearchTerm);
});

genreFilterVar.onNextChange((newGenres) => {
  sessionStorage.setItem("selectedGenres", JSON.stringify(newGenres));
});

minViewsVar.onNextChange((newMinViews) => {
  sessionStorage.setItem("minViews", String(newMinViews));
});

maxViewsVar.onNextChange((newMaxViews) => {
  sessionStorage.setItem("maxViews", String(newMaxViews));
});

sortOptionVar.onNextChange((newSortOption) => {
  sessionStorage.setItem("sortOption", newSortOption);
});

favoriteSongsVar.onNextChange((newFavorites) => {
  localStorage.setItem("favoriteSongs", JSON.stringify(newFavorites));
});

playlistsVar.onNextChange((newPlaylists) => {
  localStorage.setItem("playlists", JSON.stringify(newPlaylists));
  window.dispatchEvent(new Event("storage")); 
});

export default cache;
