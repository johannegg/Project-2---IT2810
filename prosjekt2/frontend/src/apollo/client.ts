import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
