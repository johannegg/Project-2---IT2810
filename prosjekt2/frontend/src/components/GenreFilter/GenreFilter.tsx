import "./GenreFilter.css";
import { FaFilter } from "react-icons/fa";
import { useGenreCounts } from "../../utils/hooks/useGenreCounts";
import { useReactiveVar } from "@apollo/client";
import { clearFiltersVar, genreFilterVar, maxViewsVar, minViewsVar } from "../../apollo/cache";
import { useEffect, useState, useCallback } from "react";

interface FilterProps {
	onGenreChange: (selectedGenres: string[]) => void;
	searchTerm: string;
}

export function Filter({ onGenreChange, searchTerm }: FilterProps) {
	const minViews = useReactiveVar(minViewsVar);
	const maxViews = useReactiveVar(maxViewsVar);
	const selectedGenresFromApollo = useReactiveVar(genreFilterVar);
	const clearFilters = useReactiveVar(clearFiltersVar);
	const { genreCounts, isLoading } = useGenreCounts(
		searchTerm,
		minViews,
		maxViews,
		selectedGenresFromApollo,
	);
	const [localSelectedGenres, setLocalSelectedGenres] = useState<string[]>(
		selectedGenresFromApollo || [],
	);
	const [cachedGenreCounts, setCachedGenreCounts] = useState(genreCounts);

	const handleGenreChange = useCallback(
		(genre: string) => {
			const isSelected = localSelectedGenres.includes(genre);
			const newSelectedGenres = isSelected
				? localSelectedGenres.filter((g) => g !== genre)
				: [...localSelectedGenres, genre];

			setLocalSelectedGenres(newSelectedGenres);
			genreFilterVar(newSelectedGenres);
			sessionStorage.setItem("selectedGenres", JSON.stringify(newSelectedGenres));
			onGenreChange(newSelectedGenres);
		},
		[localSelectedGenres, onGenreChange],
	);

	// Cache genre counts only when loading completes to prevent "flickering"
	useEffect(() => {
		if (!isLoading) {
			setCachedGenreCounts(genreCounts);
		}
	}, [isLoading, genreCounts]);

	// Load genres from sessionStorage on initial mount
	useEffect(() => {
		const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
		if (savedGenres.length > 0) {
			setLocalSelectedGenres(savedGenres);
			genreFilterVar(savedGenres);
		}
	}, []);

	// Reset genres when clearFilters is true
	useEffect(() => {
		if (clearFilters) {
			setLocalSelectedGenres([]);
			sessionStorage.removeItem("selectedGenres");
		}
	}, [clearFilters]);

	// Handle keyboard input for accessibility
	const handleKeyDown = useCallback(
		(event: React.KeyboardEvent<HTMLDivElement>, genre: string) => {
			if (event.key === "Enter" || event.key === " ") {
				event.preventDefault();
				handleGenreChange(genre);
			}
		},
		[handleGenreChange],
	);

	return (
		<section className="filterContainer" aria-label="Genre filter">
			<section className="filterHeader" aria-label="Filter header">
				<FaFilter className="filterSortIcon" aria-hidden="true" />
				<h2 aria-label="Genre filter section">Genre</h2>
			</section>
			<section className="categories" aria-label="Available genres">
				{cachedGenreCounts.map((genre: { name: string; count: number }) => (
					<div
						className="filterRow"
						key={genre.name}
						tabIndex={-1}
						onKeyDown={(event) => handleKeyDown(event, genre.name)}
						role="checkbox"
						aria-checked={localSelectedGenres.includes(genre.name)}
						aria-label={`Filter for genre ${genre.name} with ${genre.count} available songs`}
					>
						<input
							type="checkbox"
							id={genre.name}
							checked={localSelectedGenres.includes(genre.name)}
							onChange={() => handleGenreChange(genre.name)}
							disabled={isLoading || genre.count === 0}
							className={isLoading || genre.count === 0 ? "disabled-filter" : ""}
							tabIndex={0}
							aria-label={`Toggle ${genre.name} genre`}
						/>
						<label
							htmlFor={genre.name}
							className={genre.count === 0 ? "disabled-filter-label" : ""}
							aria-label={`Label for ${genre.name} filter`}
						>
							{genre.name.charAt(0).toUpperCase() + genre.name.slice(1)}{" "}
							<span className="filterCount">({genre.count})</span>
						</label>
					</div>
				))}
			</section>
		</section>
	);
}
