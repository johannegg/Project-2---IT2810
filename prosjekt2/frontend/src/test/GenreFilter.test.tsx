import { describe, it, expect, vi, beforeEach, Mock } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { genreFilterVar, minViewsVar, maxViewsVar } from "../apollo/cache";
import { MockedProvider } from "@apollo/client/testing";
import { Filter } from "../components/GenreFilter/GenreFilter";

const mockUseGenreCounts = vi.fn();

vi.mock("../utils/hooks/useGenreCounts", () => ({
  useGenreCounts: () => mockUseGenreCounts(),
}));

vi.mock("@apollo/client", async () => {
  const actual = await vi.importActual("@apollo/client");
  return {
    ...actual,
    useReactiveVar: (variable: any) => variable(),
  };
});

describe("Filter Component", () => {
  let onGenreChangeMock: Mock<(selectedGenres: string[]) => void>;

  beforeEach(() => {
    onGenreChangeMock = vi.fn();
    genreFilterVar([]);
    minViewsVar(0);
    maxViewsVar(1000);
    mockUseGenreCounts.mockReset();
  });

  it("does not render checkboxes while loading", () => {
    mockUseGenreCounts.mockImplementation(() => ({
      genreCounts: [],
      isLoading: true,
      error: undefined,
      refetch: vi.fn(),
    }));

    render(
      <MockedProvider>
        <Filter onGenreChange={onGenreChangeMock} searchTerm="" />
      </MockedProvider>
    );

    const countryCheckbox = screen.queryByLabelText(/Country/);
    expect(countryCheckbox).not.toBeInTheDocument();
  });

  it("renders genres correctly", async () => {
    mockUseGenreCounts.mockImplementation(() => ({
      genreCounts: [
        { name: "Country", count: 59 },
        { name: "Pop", count: 990 },
        { name: "Rap", count: 1246 },
        { name: "Rb", count: 147 },
        { name: "Rock", count: 505 },
      ],
      isLoading: false,
      error: undefined,
      refetch: vi.fn(),
    }));

    render(
      <MockedProvider>
        <Filter onGenreChange={onGenreChangeMock} searchTerm="example" />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(screen.getByText(/Country/)).toBeInTheDocument();
      expect(screen.getByText(/Pop/)).toBeInTheDocument();
      expect(screen.getByText(/Rap/)).toBeInTheDocument();
      expect(screen.getByText(/Rb/)).toBeInTheDocument();
      expect(screen.getByText(/Rock/)).toBeInTheDocument();
    });
  });
});
