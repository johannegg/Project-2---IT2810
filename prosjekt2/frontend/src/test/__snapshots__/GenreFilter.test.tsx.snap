// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Filter Component > renders genres correctly and matches the snapshot 1`] = `
<div>
  <section
    aria-label="Genre filter"
    class="filterContainer"
  >
    <section
      aria-label="Filter header"
      class="filterHeader"
    >
      <svg
        aria-hidden="true"
        class="filterSortIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
        />
      </svg>
      <h2
        aria-label="Genre filter section"
      >
        Genre
      </h2>
    </section>
    <section
      aria-label="Available genres"
      class="categories"
    >
      <div
        aria-checked="false"
        aria-label="Filter for genre Country with 59 available songs"
        class="filterRow"
        role="checkbox"
        tabindex="-1"
      >
        <input
          aria-label="Toggle Country genre"
          class=""
          id="Country"
          tabindex="0"
          type="checkbox"
        />
        <label
          aria-label="Label for Country filter"
          class=""
          for="Country"
        >
          Country
           
          <span
            class="filterCount"
          >
            (
            59
            )
          </span>
        </label>
      </div>
      <div
        aria-checked="false"
        aria-label="Filter for genre Pop with 990 available songs"
        class="filterRow"
        role="checkbox"
        tabindex="-1"
      >
        <input
          aria-label="Toggle Pop genre"
          class=""
          id="Pop"
          tabindex="0"
          type="checkbox"
        />
        <label
          aria-label="Label for Pop filter"
          class=""
          for="Pop"
        >
          Pop
           
          <span
            class="filterCount"
          >
            (
            990
            )
          </span>
        </label>
      </div>
      <div
        aria-checked="false"
        aria-label="Filter for genre Rap with 1246 available songs"
        class="filterRow"
        role="checkbox"
        tabindex="-1"
      >
        <input
          aria-label="Toggle Rap genre"
          class=""
          id="Rap"
          tabindex="0"
          type="checkbox"
        />
        <label
          aria-label="Label for Rap filter"
          class=""
          for="Rap"
        >
          Rap
           
          <span
            class="filterCount"
          >
            (
            1246
            )
          </span>
        </label>
      </div>
      <div
        aria-checked="false"
        aria-label="Filter for genre Rb with 147 available songs"
        class="filterRow"
        role="checkbox"
        tabindex="-1"
      >
        <input
          aria-label="Toggle Rb genre"
          class=""
          id="Rb"
          tabindex="0"
          type="checkbox"
        />
        <label
          aria-label="Label for Rb filter"
          class=""
          for="Rb"
        >
          Rb
           
          <span
            class="filterCount"
          >
            (
            147
            )
          </span>
        </label>
      </div>
      <div
        aria-checked="false"
        aria-label="Filter for genre Rock with 505 available songs"
        class="filterRow"
        role="checkbox"
        tabindex="-1"
      >
        <input
          aria-label="Toggle Rock genre"
          class=""
          id="Rock"
          tabindex="0"
          type="checkbox"
        />
        <label
          aria-label="Label for Rock filter"
          class=""
          for="Rock"
        >
          Rock
           
          <span
            class="filterCount"
          >
            (
            505
            )
          </span>
        </label>
      </div>
    </section>
  </section>
</div>
`;
